### PR TITLE
Improve docs on Arc<T> and Send/Sync

### PR DESCRIPTION
**Pull request by steveklabnik**
This is something I always forget, so let's actually
explain in the docs.

I didn't fully link up everything here, but I'd like to make sure that the wording is okay before I bother.

### CODE REVIEW

File: src/liballoc/arc.rs [hunk 1]
```diff
@@ -54,16 +54,32 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
...
+/// are not sharing reference-counted values between threads, consider using
+/// [`Rc<T>`] for lower overhead. [`Rc<T>`] is a safe default, because the
+/// compiler will catch any attempt to send an [`Rc<T>`] between threads.
+/// However, a library might choose `Arc<T>` in order to give library consumers
 /// more flexibility.
 ///
+/// `Arc<T>` will implement [`Send`] and [`Sync`] as long as the `T` implements
+/// [`Send`] and [`Sync`]. This may be a bit counter-intuitive at first: after
+/// all, isn't the point of `Arc<T>` thread safety? The key is this: `Arc<T>`
+/// is an implementation of multiple ownership that is thread safe, but it
```
Comment (oli-obk) [inline src/liballoc/arc.rs]: Maybe state here, that `Arc<T>` is threadsafe in the way that you can safely clone an `Arc<T>` pointing to the same data in multiple threads at the same time.

File: src/liballoc/arc.rs [hunk 2]
```diff
@@ -54,16 +54,32 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
...
+/// atomic operations are more expensive than ordinary memory accesses.  If you
+/// are not sharing reference-counted values between threads, consider using
+/// [`Rc<T>`] for lower overhead. [`Rc<T>`] is a safe default, because the
+/// compiler will catch any attempt to send an [`Rc<T>`] between threads.
+/// However, a library might choose `Arc<T>` in order to give library consumers
 /// more flexibility.
 ///
+/// `Arc<T>` will implement [`Send`] and [`Sync`] as long as the `T` implements
+/// [`Send`] and [`Sync`]. This may be a bit counter-intuitive at first: after
+/// all, isn't the point of `Arc<T>` thread safety? The key is this: `Arc<T>`
```
Comment (kennytm) [inline src/liballoc/arc.rs]: I find that saying "`T: Send/Sync` ⇔ `Arc<T>: Send/Sync` is counter-intuitive" is counter-intuitive 😅. It may be clearer to add one more sentence explaining the counter-intuitiveness, e.g. "Why it is not able hide the non-thread-safe type `T` to make itself `Arc<T>` thread-safe?"

File: src/liballoc/arc.rs [hunk 3]
```diff
@@ -54,16 +54,32 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
...
-/// that atomic operations are more expensive than ordinary memory accesses.
-/// If you are not sharing reference-counted values between threads, consider
-/// using [`rc::Rc`][`Rc`] for lower overhead. [`Rc`] is a safe default, because
-/// the compiler will catch any attempt to send an [`Rc`] between threads.
-/// However, a library might choose `Arc` in order to give library consumers
+/// ## Thread Safety
+///
+/// Unlike [`Rc<T>`], `Arc<T>` uses atomic operations for its reference
+/// counting.  This means that it is thread-safe. The disadvantage is that
+/// atomic operations are more expensive than ordinary memory accesses.  If you
```
Comment (killercup) [inline src/liballoc/arc.rs]: The most tiniest of nits: What's with the double spaces after _some_ sentences? 😄

**Comment by steveklabnik** [comment]
tl;dr: a bad vim config, and me missing fixes in some spots. ha!

File: src/liballoc/arc.rs (hunk 1 above)
Comment (steveklabnik) [inline src/liballoc/arc.rs]: How would you construct this if you cannot send an arc to a different thread?

Comment (oli-obk) [inline src/liballoc/arc.rs]: If you had a type `T` that is `!Sync` and an `Arc<T>` that does not give you access to the internal object, then the `Arc<T>` can be `Send`, because you can never access the inner data, but you can clone and drop the `Arc<T>` to your heart's desire. This is the guarantee that `Arc<T>` gives, without looking at the `T`, you can clone and drop the `Arc<T>`. Since `Arc<T>` also gives you access to it's inner data through `Deref`, you also need the `Sync` bound on `T` to make `Arc<T>` `Send`, but that's nothing special about `Arc<T>` but a general thing about sharing data over thread boundaries as stated in the `Sync` docs.

Comment (steveklabnik) [inline src/liballoc/arc.rs]: but `impl<T> Send for Arc<T> where T: Send + Sync + ?Sized`, that is, if `T` is `!Sync`, the `Arc<T>` will not be `Send`.

Comment (oli-obk) [inline src/liballoc/arc.rs]: I understand that. I just meant to explain the reasoning behind that a bit in different words than "is an implementation of multiple ownership that is thread safe", because i totally stumbled over that sentence and would have not understood it at all 5 years ago

Comment (steveklabnik) [inline src/liballoc/arc.rs]: ah okay. I wish I had a way to re-word it that didn't involve something that's not true....... hm

Comment (oli-obk) [inline src/liballoc/arc.rs]: Maybe something like

> `Arc<T>` makes it thread safe to create multiple pointers to the same data, but it ...

**Comment by steveklabnik** [comment]
I plan on updating this PR early next week; sorry about the delay.

File: src/liballoc/arc.rs [hunk 4]
```diff
@@ -54,16 +54,32 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
...
 ///
+/// `Arc<T>` will implement [`Send`] and [`Sync`] as long as the `T` implements
+/// [`Send`] and [`Sync`]. This may be a bit counter-intuitive at first: after
+/// all, isn't the point of `Arc<T>` thread safety? The key is this: `Arc<T>`
+/// is an implementation of multiple ownership that is thread safe, but it
+/// doesn't add thread safety to its data. Consider `Arc<RefCell<T>>`.
+/// `RefCell<T>` isn't [`Send`], and if `Arc<T>` was always [`Send`],
+/// `Arc<RefCell<T>>` would be as well. But then we'd have a problem:
+/// `RefCell<T>` is not thread safe; it keeps track of the borrowing count
+/// using non-atomic operations. It works the same way with [`Sync`].
```
Comment (BurntSushi) [inline src/liballoc/arc.rs]: `RefCell<T>` is `Send` when `T` is `Send`. :-) I think you could either swap `RefCell<T>` for `Rc<T>` and change the "borrowing count" to "reference count," or you could keep `RefCell<T>` and change `Send` to `Sync` (since `RefCell<T>` is never `Sync`) and then drop the last sentence.

File: src/liballoc/arc.rs [hunk 5]
```diff
@@ -54,16 +54,32 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
...
+///
+/// In the end, this means that you may need to pair `Arc<T>` with some sort of
+/// `std::sync` type, usually `Mutex<T>`.
```
Comment (BurntSushi) [inline src/liballoc/arc.rs]: Yeah, this is an important sentence, and it probably flows better if you keep `RefCell` in the preceding paragraph.

**Comment by BurntSushi** [comment]
I like this @steveklabnik. :-) r=me once feedback is addressed.

File: src/liballoc/arc.rs (hunk 1 above)
Comment (steveklabnik) [inline src/liballoc/arc.rs]: I've tweaked this slightly, let me know what you think.

**Comment by steveklabnik** [comment]
I believe I've fixed up everything, but I'd like to hear what you all have to say before I use @BurntSushi 's `r=me` 😄

**Comment by kennytm** [comment]
LGTM :+1:. There is still an instance of double-space but I don't think it is a big issue.

**Comment by BurntSushi** [comment]
LGTM.
